reqwest-middleware/src/middleware.rs

```diff
@@ -46,7 +47,8 @@ pub trait Middleware: 'static + Send + Sync {
     ) -> Result<Response>;
 }
 
-#[async_trait::async_trait]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
```

Comment:
I think I would like to find a way to avoid this feature dance. Currently not too sure how just yet.

I did experiment with making a more custom trait without async_trait where the future is an associated type. This makes the Send bound potentially unnecessary.

Reply:
This would necessarily be a large structural change. However with a jump to http::Extensions that is also a breaking change

Reply:
I think having the future being an associated type would have the same issue: either you still force the associated type to be Send, and it won't work on wasm, or you don't force the associated type to be Send, and then non-wasm versions can no longer spawn that future onto a task-stealing executor without introspection into what middleware it exactly is.

FWIW this is the solution suggested at https://github.com/dtolnay/async-trait/issues/159. Though I kinda would have liked async_trait to have this default upstream, I'm not really expecting it to happen as it'd be a breaking change of non-obvious value (code starts behaving differently between wasm and non-wasm, so it's not a no-brainer) and async-trait seems quite stable at its 0.1 version.

A third solution would be to PR reqwest upstream to make its futures Send, but given that wasm doesn't have threads I don't have much hope of it happening either…

For the switch to http::Extensions, it'd also be possible to do the switch to http::Extensions only on wasm so that it's not a breaking change (this way only a new platform is supported), but such a significant variation across platforms would feel wrong to me ^^'

WDYT?

Reply:
> I think having the future being an associated type would have the same issue: either you still force the associated type to be Send

Actually this isn't true. We can avoid having the send bound and the resulting future can still happen to be Send.

We don't explicitly use tokio::spawn_blocking anywhere or need Send otherwise, so it should technically be ok.

Reply:
However, last time I tried it I got very upset that reqwest-middleware makes its internal request future private. Which necessitated Boxing it. Though we can just have the cfg on that specific future to be `Box<dyn Future + ?Send>`

Reply:
reqwest-middleware doesn't use tokio::spawn, but consumers of the crate do require tokio::spawn. Even if the resulting future is Send, at the time the first Middleware is trait-objectified, the compiler will stop knowing that the future below it is actually Send. And then, when the user will call execute, it'll stop working.

One way to work around this would be to replace ClientWithMiddleware to have a generic list of middlewares so that middlewares are never dyn'ified, but it probably would start being too much type engineering to be worth it.

As for not-boxing, it's hopefully coming soon, as I saw movement on the type-alias-impl-trait front just today IIRC.